src/plugins/map/MapObject.cpp

```diff
@@ -1260,67 +1281,99 @@ void MapObjectMapLoc::UpdateText()
 
 //----------------------------------------------------------------------------
 
-MapObjectMapLoc* GetMapLocByIndex(size_t index)
+MapObjectMapLoc* MapLocTemplate::GetMapLoc()
 {
-	--index;
-
-	if (index >= sMapLocs.size())
-		return nullptr;
-
-	return sMapLocs[index];
+	return m_MapLoc;
 }
 
-MapObjectMapLoc* GetMapLocByTag(const std::string& tag)
+MapLocTemplate* GetMapLocByTag(const std::string& tag)
 {
-	auto iter = sLocationsMap.find(tag);
-	if (iter == sLocationsMap.end())
-		return nullptr;
+	for (auto maplocTemplate : gMapLocTemplates)
+	{
+		if (!strcmp(maplocTemplate->GetMapLoc()->GetTag().c_str(), tag.c_str()))
+			return maplocTemplate;
+	}
 
-	return iter->second;
+	return nullptr;
 }
 
 void DeleteAllMapLocs()
 {
-	sLocationsMap.clear();
-
-	for (MapObjectMapLoc* loc : sMapLocs)
+	for (auto mapLocTemplate : gMapLocTemplates)
 	{
-		delete loc;
+		delete mapLocTemplate->GetMapLoc();
+		mapLocTemplate->ClearReferenceToMapLoc();
 	}
 
-	sMapLocs.clear();
+	gMapLocTemplates.clear();
 }
 
 void UpdateMapLocIndexes()
 {
-	for (int i = 0; i < (int)sMapLocs.size(); ++i)
+	for (size_t i = 0; i < gMapLocTemplates.size(); ++i)
 	{
-		sMapLocs[i]->SetIndex(i + 1);
+		gMapLocTemplates[i]->GetMapLoc()->SetIndex(i + 1);
 	}
 }
 
 void DeleteMapLoc(MapObjectMapLoc* mapLoc)
 {
-	sMapLocs.erase(
-		std::remove(sMapLocs.begin(), sMapLocs.end(), mapLoc),
-		sMapLocs.end());
-	sLocationsMap.erase(mapLoc->GetTag());
+	gMapLocTemplates.erase(std::remove_if(gMapLocTemplates.begin(), gMapLocTemplates.end(),
+		[mapLoc](MapLocTemplate* match) -> bool
+		{
+			if (match->GetMapLoc()->GetTag() == mapLoc->GetTag())
+			{
+				delete match->GetMapLoc();
+				match->ClearReferenceToMapLoc();
+				return true;
```

Comment:
If MapLoc's destructor called ClearReferenceToMapLoc, and MapLocTemplate's destructor deleted MapLoc, then you don't need these statements. just return true

Reply:
so I made these changes, but it now throws in the mapobjectmaploc dtor when trying to look up the template to deref which is mid deletion as well.  I can maybe work around this but there's more going on than just these tweaks I think